File: serdaSimus/assignments/04mergeIntervals/lc56/lc56.js
// Problem Statement #

// Given a list of intervals, merge all the overlapping intervals to produce a list that has only mutually exclusive intervals.

class Interval {
  constructor(start, end) {
    this.start = start;
    this.end = end;
  }

  get_interval() {
    return "[" + this.start + ", " + this.end + "]";
  }
}

const merge = function (intervals) {
  merged = [];
  // TODO: Write your code here
  //edge: 1 item, no item.
  if (intervals.length == 1 || !intervals.length) return intervals;

  // O(nlogn) sort
  intervals.sort((a, b) => a.start - b.start);

  // assign the first interval as previous:
  merged.push(intervals[0]);

  // visit all subsequent intervals to find/merge overlapping intervals.
  for (let index = 1; index < intervals.length; index++) {
    let previousPtr = merged.pop();

    // point the currentPtr to the element of intervals at the current index of the loop.
    let currentPtr = intervals[index];

    //comparisons: is previous last index, more than current's first index.
    if (previousPtr.end < currentPtr.start) {
      merged.push(previousPtr);
      merged.push(currentPtr);
      // inserting the last item

      continue;
    } else {
      if (previousPtr.end < currentPtr.end) {
        // push into merge [prv.start, current.end]
        merged.push(new Interval(previousPtr.start, currentPtr.end));
      } else {
        // meaning previous end is more than current.end, previous starts earlier, ends later.
        merged.push(previousPtr);
        continue;
      }
    }
  }
  return merged;
};

merged_intervals = merge([
  new Interval(1, 4),
  new Interval(2, 5),
  new Interval(7, 9),
]);
result = "";
for (i = 0; i < merged_intervals.length; i++) {
  result += merged_intervals[i].get_interval() + " ";
}
console.log(`Merged intervals: ${result}`);

merged_intervals = merge([
  new Interval(6, 7),
  new Interval(2, 4),
  new Interval(5, 9),
]);
result = "";
for (i = 0; i < merged_intervals.length; i++) {
  result += merged_intervals[i].get_interval() + " ";
}
console.log(`Merged intervals: ${result}`);

merged_intervals = merge([
  new Interval(1, 4),
  new Interval(2, 6),
  new Interval(3, 5),
]);
result = "";
for (i = 0; i < merged_intervals.length; i++) {
  result += merged_intervals[i].get_interval() + " ";
}
console.log(`Merged intervals: ${result}`);

// Solution
// -----
// function merge(intervals) {
//   if (intervals.length < 2) {
//     return intervals;
//   }
//   // sort the intervals on the start time
//   intervals.sort((a, b) => a.start - b.start);

//   const mergedIntervals = [];
//   let start = intervals[0].start,
//     end = intervals[0].end;
//   for (i = 1; i < intervals.length; i++) {
//     const interval = intervals[i];
//     if (interval.start <= end) { // overlapping intervals, adjust the 'end'
//       end = Math.max(interval.end, end);
//     } else { // non-overlapping interval, add the previous interval and reset
//       mergedIntervals.push(new Interval(start, end));
//       start = interval.start;
//       end = interval.end;
//     }
//   }
//   // add the last interval
//   mergedIntervals.push(new Interval(start, end));
//   return mergedIntervals;
// }

// -----

// Time complexity #
// The time complexity of the above algorithm is O(N * logN), where ‘N’ is the total number of intervals. We are iterating the intervals only once which will take O(N), in the beginning though, since we need to sort the intervals, our algorithm will take O(N * logN).

// Space complexity #
// The space complexity of the above algorithm will be O(N) as we need to return a list containing all the merged intervals. We will also need O(N) space for sorting. For Java, depending on its version, Collections.sort() either uses Merge sort or Timsort, and both these algorithms need O(N) space. Overall, our algorithm has a space complexity of O(N).

File: AdrianBarros/assignments/mergeIntervals/lc56/lc56.py
# Problem Statement

# Given a list of intervals, merge all the overlapping intervals to produce a list that has only mutually exclusive intervals.

from __future__ import print_function


class Interval:
    def __init__(self, start, end):
        self.start = start
        self.end = end

    def print_interval(self):
        print("[" + str(self.start) + ", " + str(self.end) + "]", end='')


'''
input: array of pairs/intervals
output: array of merged intervals
step 1: sort the array using the first number of the interval
step 2: loop through the array initializing a copy of the first pair
    step 3: conditional
         IF curr_second > next_first: merge intervals [curr_first, max(next_second,curr_second)]
         ELSE: append current pair
###after thoughts we need to add the last interval at the end
FINAL return merged[]
'''


def merge(intervals):
    merged = []
    # TODO: Write your code here
    # step 1: sort the array using the first number of the interval
    intervals.sort(key=lambda x: x.start)

    # step 2: loop through the array initializing a copy of the first pair
    merged_int = intervals[0]  # decided to change the name from curr to merged
    for i in range(1, len(intervals)):  # starts from 1 bc we are using 0

        # step 3: conditional
        curr_int = intervals[i]
        if merged_int.end > curr_int.start:
            merged_int.end = max(merged_int.end, curr_int.end)
        else:
            merged.append(merged_int)
            merged_int = curr_int
    merged.append(merged_int)
    return merged


def main():
    print("Merged intervals: ", end='')
    for i in merge([Interval(1, 4), Interval(2, 5), Interval(7, 9)]):
        i.print_interval()
    print()

    print("Merged intervals: ", end='')
    for i in merge([Interval(6, 7), Interval(2, 4), Interval(5, 9)]):
        i.print_interval()
    print()

    print("Merged intervals: ", end='')
    for i in merge([Interval(1, 4), Interval(2, 6), Interval(3, 5)]):
        i.print_interval()
    print()


main()


# Solution
# -----
# def merge(intervals):
#   if len(intervals) < 2:
#     return intervals

#   # sort the intervals on the start time
#   intervals.sort(key=lambda x: x.start)

#   mergedIntervals = []
#   start = intervals[0].start
#   end = intervals[0].end
#   for i in range(1, len(intervals)):
#     interval = intervals[i]
#     if interval.start <= end:  # overlapping intervals, adjust the 'end'
#       end = max(interval.end, end)
#     else:  # non-overlapping interval, add the previous internval and reset
#       mergedIntervals.append(Interval(start, end))
#       start = interval.start
#       end = interval.end

#   # add the last interval
#   mergedIntervals.append(Interval(start, end))
#   return mergedIntervals

# -----
